task/control-plane-agent/src/inventory.rs

```diff
@@ -22,32 +30,65 @@ impl Inventory {
 
         Self {
             validate_task: Validate::from(VALIDATE.get_task_id()),
+            sensor_task: SensorTask::from(SENSOR.get_task_id()),
         }
     }
 
     pub(crate) fn num_devices(&self) -> usize {
         OUR_DEVICES.len() + VALIDATE_DEVICES.len()
     }
 
-    pub(crate) fn device_description(
+    pub(crate) fn num_component_details(
         &self,
-        index: usize,
-    ) -> DeviceDescription<'static> {
-        // If `index` is in `0..OUR_DEVICES.len()`, return that device directly;
-        // otherwise, subtract `OUR_DEVICES.len()` to shift it into the range
-        // `0..VALIDATE_DEVICES.len()` and ask `validate`.
-        if index < OUR_DEVICES.len() {
-            OUR_DEVICES[index]
-        } else {
-            let index = index - OUR_DEVICES.len();
-            self.device_description_for_validate_device(index)
+        component: &SpComponent,
+    ) -> Result<u32, SpError> {
+        match Index::try_from(component)? {
+            Index::OurDevice(_) => Ok(0),
+            Index::ValidateDevice(i) => {
+                Ok(VALIDATE_DEVICES[i].sensors.len() as u32)
+            }
         }
     }
 
-    fn device_description_for_validate_device(
+    pub(crate) fn component_details(
+        &self,
+        component: &SpComponent,
+        component_index: u32,
+    ) -> ComponentDetails {
+        // We should only be called if we returned `Ok(n)` with `n > 0` from
+        // `num_component_details()` with this same component, so we should only
+        // be called if (a) `Index::try_from(component)` succeeded and (b) the
+        // returned value was an `Index::ValidateDevice(i)` (with at least i+1
+        // sensors). We panic if we're called incorrectly.
+        let val_device_index = match Index::try_from(component) {
+            Ok(Index::ValidateDevice(i)) => i,
+            Ok(Index::OurDevice(_)) | Err(_) => panic!(),
+        };
+
+        let sensor_description = &VALIDATE_DEVICES[val_device_index].sensors
+            [component_index as usize];
+
+        let value = self
+            .sensor_task
+            .get(sensor_description.id)
+            .map_err(|err| SensorErrorConvert(err).into());
+
+        ComponentDetails::Measurement(Measurement {
+            name: sensor_description.name.unwrap_or(""),
+            kind: MeasurementKindConvert(sensor_description.kind).into(),
+            value,
+        })
+    }
+
+    pub(crate) fn device_description(
         &self,
         index: usize,
     ) -> DeviceDescription<'static> {
+        let index = match Index::from_overall_index(index) {
```

Comment:
Is `index` trusted at this point?  Otherwise, it could panic the function here.

Reply:
It is; I'll add comments explaining.

Reply:
Added comments and a `BoundsChecked` newtype in 73088dc3